src/neuroglancer/ui/segment_list.ts

```diff
@@ -82,34 +87,38 @@ class SegmentListSource extends RefCounted implements VirtualListSource {
     const splices: ArraySpliceOp[] = [];
     let changed = false;
     let matchStatusTextPrefix = '';
-    const {visibleSegments} = this.segmentationDisplayState.segmentationGroupState.value;
+    const {selectedSegments, visibleSegments} = this.segmentationDisplayState.segmentationGroupState.value;
     const visibleSegmentsGeneration = visibleSegments.changed.count;
+    const selectedSegmentsGeneration = selectedSegments.changed.count;
     const prevVisibleSegmentsGeneration = this.visibleSegmentsGeneration;
+    const prevSelectedSegmentsGeneration = this.selectedSegmentsGeneration;
+    const segmentsChanged = prevVisibleSegmentsGeneration !== visibleSegmentsGeneration ||
+      prevSelectedSegmentsGeneration !== selectedSegmentsGeneration
     const unconstrained = isQueryUnconstrained(queryResult.query);
-    if (unconstrained) {
+    if (unconstrained && this.showExplicit) {
       // Full list of visible segments is shown only if no query is specified.
-      if (prevVisibleSegmentsGeneration !== visibleSegmentsGeneration ||
+      if (segmentsChanged ||
           this.explicitSegments === undefined || !this.explicitSegmentsVisible) {
         this.visibleSegmentsGeneration = visibleSegmentsGeneration;
-        const newSortedVisibleSegments = Array.from(visibleSegments, x => x.clone());
-        newSortedVisibleSegments.sort(Uint64.compare);
+        const newSelectedSegments = [...selectedSegments];
         const {explicitSegments} = this;
         if (explicitSegments === undefined) {
-          this.explicitSegments = newSortedVisibleSegments;
+          this.explicitSegments = newSelectedSegments;
           splices.push(
-              {retainCount: 0, insertCount: newSortedVisibleSegments.length, deleteCount: 0});
+              {retainCount: 0, insertCount: newSelectedSegments.length, deleteCount: 0});
         } else {
           splices.push(
-              ...getMergeSplices(explicitSegments, newSortedVisibleSegments, Uint64.compare));
+              ...getMergeSplices(explicitSegments, newSelectedSegments, Uint64.compare));
         }
-        this.explicitSegments = newSortedVisibleSegments;
+        this.explicitSegments = newSelectedSegments;
         changed = true;
       } else {
         splices.push({retainCount: this.explicitSegments.length, deleteCount: 0, insertCount: 0});
       }
       this.explicitSegmentsVisible = true;
```

Comment:
github won't let me tag line 132 but what is the reason for `this.explicitSegments = explicitIds;` from the query result since it seems explicitSegmentsVisible is always false if there is a query.

Reply:
The way it is currently implemented, is that there are two parts of the list, the `explicitSegments` part and the matches part.  The `explicitSegments`  lists the segment ids directly, while the matches list the segments by their index into the segment property map.  The `explicitSegments` part is used for two purposes:

1. If the query is unconstrained, then for showing the visible segments.
2. If the user specifies explicit segment ids as the query, i.e. `12345, 6789`.  These have to be in the `explicitSegments` portion because there is not necessarily an entry in the segment property map for these ids.